XenModel/PoolJoinRules.cs

```diff
@@ -388,6 +397,35 @@ private static bool DifferentServerVersion(Host slave, Host master)
                 slave.ProductBrand != master.ProductBrand;
         }
 
+        /// <summary>
+        /// Check whether all updates that request homogeneity are in fact homogeneous
+        /// across the proposed pool. This is used in CanJoinPool and prevents the pool from being created
+        /// </summary>
+        private static bool DifferentHomogeneousUpdates(Host slave, Host master)
+        {
+            if (slave == null || master == null)
+                return false;
+
+            List<Host> allHosts = new List<Host>(master.Connection.Cache.Hosts);
+            allHosts.Add(slave);
+
+            // Make a list of updates that should be homogeneous
+            var homogeneousUpdates = new List<Pool_update>();
+            foreach (var host in allHosts)
+            {
+                homogeneousUpdates.AddRange(host.AppliedUpdates().Where(update => update.enforce_homogeneity));
+            }
+
+            foreach (var update in homogeneousUpdates)
+            {
+                if (allHosts.Any(host => !host.AppliedUpdates().Exists(u => u.uuid == update.uuid)))

```

Comment:
I think @kc284's check isn't correct, but there should be a more efficient way to do it. For example, construct a Map<Update, int> showing the number of hosts having each update: for each update, that number should be the same as the number of hosts in the pool.